### PR TITLE
AddTransaction component has functionality to update state and backend API

### PR DESCRIPTION
**Pull request by NicolajNiels1**
- [x] Choice of Expense or Income should be done with a radio button
- [x] The AddTransaction component should submit a form
- [x] A new entry should be posted to the API
- [x] When a transaction is posted to database, the transaction state is updated.
- [x] A new entry should be added to the Transaction list

_This is a rather large PR. Sorry for that. There's a lot of minor changes, like code formatting and moved folders that touches many files. In the future I'll be careful not to include these kinds of changes which will result in much smaller PR._ 

I will happily sit down with @KaizerTilt and @FredrikAugust and run through the PR

### Summary
In this PR I have added functionality for all of the items in the checklist above.

Additionally, I have written a `TransactionProvider` component in `src/store/contexts/transactions/` which takes care of most of the logic with React Context API and related hooks (useContext, useReducer). It should be 
- easier to use, less to think about for the developer
- all logic gathered one place, in stead of everywhere
- More explicit
- More optimal rendering
- Easier for testing individual components (without checking if context exists, creating the context etc...)

Moreover,
-  I have moved `reducer/` and `context/` folder to `src/store`
- I have created a `TransactionMocker` component in `src/helpers/create_transactions.tsx` which can be used as a child-component of `TransactionProvider` for easily faking transaction data in the state. This should become obsolete once we have a more stable dev server up and running. 
- `Nock` is a new dependency I have added to dev. This makes testing the api client much faster. See [documentation here](https://github.com/nock/nock)

### CODE REVIEW

File: src/components/atoms/__tests__/TransactionEntry.test.tsx [hunk 1]
```diff
@@ -1,46 +1,31 @@
...
         date={testDate}
-        companyId={testCompanyId}
+        company_id={testCompanyId}
         type={testType}
         notes={testNotes}
-      />
-    </Wrapper>
-  ));
+    />
+  ), {wrapper: TransactionProvider});
```
Comment (NicolajNiels1) [inline src/components/atoms/__tests__/TransactionEntry.test.tsx]: This is an easier way to wrap tests.

File: src/components/molecules/AddTransaction.tsx [hunk 2]
```diff
@@ -38,52 +84,67 @@ const AddTransaction: React.FC<IProps> = props => {
...
+            checked={transactionType === 'expense'}
+            onChange={handleTransactionTypeChange}
+          />{' '}
+          Expense
+          <input
+            type="radio"
+            name="transactionType"
+            value={'income'}
+            onChange={handleTransactionTypeChange}
+          />{' '}
```
Comment (NicolajNiels1) [inline src/components/molecules/AddTransaction.tsx]: These radio buttons should be designed more carefully in another PR

File: src/components/molecules/TransactionCalculator.tsx [hunk 3]
```diff
@@ -8,18 +8,17 @@ import React from 'react';
...
+import { useTransactionState } from '../../store/contexts/transactions';
 import TransactionEntry from '../atoms/TransactionEntry';
 
 interface ITransactionCalculatorProps {
   className?: string;
 }
 
 const TransactionCalculator: React.FC<ITransactionCalculatorProps> = props => {
-  const { store } = React.useContext(TransactionCtx);
+  const store = useTransactionState();
```
Comment (NicolajNiels1) [inline src/components/molecules/TransactionCalculator.tsx]: More human-readable 👀

File: src/components/organism/Balances.tsx [hunk 4]
```diff
@@ -12,7 +12,8 @@ interface IProps {
   className?: string;
 }
 
-const companyId = 1; // Hardcoded until we get a global company context.
+// tslint:disable-next-line: variable-name
+const company_id = 1; // Hardcoded until we get a global company context.
```
Comment (NicolajNiels1) [inline src/components/organism/Balances.tsx]: Comply with change in ITransaction which complies with API response

File: src/helpers/GlobalWrapper.tsx [hunk 5]
```diff
@@ -30,34 +28,24 @@ interface IWrapperProps {
...
-   * If we haven't initialized the context, we want to create it here.
-   */
-  if (!TransactionCtx) {
-    createTransactionCtx(store, dispatch);
-
-    for (let i = 0; i < 100; i++) {
-      dispatch(ActionCreators.addTransaction(createDummyTransaction()));
-    }
-  }
-
```
Comment (NicolajNiels1) [inline src/helpers/GlobalWrapper.tsx]: All of this logic is gathered in `src/store/contexts/transactions`

File: src/mitochondria/transactions.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ * @param authState The authentication state for authenticating with the API
+ */
+export const createTransaction = async (transaction: INewTransaction, authState: IAuthState) => {
+  try {
+    const response = await post('/transaction', transaction, authState);
+    return response;
+  } catch (e) {
+    throw e;
+  }
+};
```
Comment (NicolajNiels1) [inline src/mitochondria/transactions.ts]: We should discuss whether or not this kind of specialization is necessary.

File: src/store/contexts/transactions.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+};
+
+// Exports
+export {
+  initialState,
+  TransactionProvider,
+  useTransactions,
+  useTransactionState,
+  useTransactionDispatch,
+};
```
Comment (NicolajNiels1) [inline src/store/contexts/transactions.tsx]: I really like this way of using the useContext, useReducer hooks with Reacts Context API,
see the article about [How to use react context effectively](https://kentcdodds.com/blog/how-to-use-react-context-effectively)

**Comment by FredrikAugust** [comment]
Sounds good, would love to go through it. I'll be back at work on Monday.

File: src/mitochondria/index.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+export const post = async (
+  endpoint: ApiEndpoint,
+  data: { [key: string]: any },
+  authState: IAuthState,
+) => {
+  const url = new URL(`${BASE_URL}${endpoint}/`);
+  const initOptions = {
+    body: JSON.stringify(data),
+    headers: {
+      'Authentication': `Bearer ${authState.access}`,
```
Comment (FredrikAugust) [inline src/mitochondria/index.ts]: localStorage.getItem('access');

File: src/mitochondria/index.ts [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+ * anything else that prevented the fetch() request to be completed.
+ * TODO – This method should be develped more to handle errors in a more unified way throughout the application.
+ */
+const handleError = (error: Error) => {
+  return new Error(error.message);
+};
+
+// Re-export everything from mitochondria for easier imports
+export * from './balances';
+export * from './transactions';
```
Comment (FredrikAugust) [inline src/mitochondria/index.ts]: Smart! :thinking:

File: src/mitochondria/transactions.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import { TransactionType } from '../declarations/transaction';
+import { IAuthState } from '../store/reducers/auth';
+import {post} from './';
+
+/**
+ * Interface for body of a post request (Add Transaction) of the API
+ */
+export interface INewTransaction {
```
Comment (FredrikAugust) [inline src/mitochondria/transactions.ts]: `Omit<ITransaction, 'id'>`

File: src/mitochondria/transactions.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+ * @param transaction The transaction we want to post to the API
+ * @param authState The authentication state for authenticating with the API
+ */
+export const createTransaction = async (transaction: INewTransaction, authState: IAuthState) => {
+  try {
+    const response = await post('/transaction', transaction, authState);
+    return response;
+  } catch (e) {
+    throw e;
+  }
```
Comment (FredrikAugust) [inline src/mitochondria/transactions.ts]: Change to `await post(...)` as it will throw if it goes wrong anyway.

File: src/components/molecules/AddTransaction.tsx (hunk 2 above)
Comment (FredrikAugust) [reply src/components/molecules/AddTransaction.tsx]: I also don't think you need to type `{'income'}`. You should be able to type just `"income"`, which I believe should be a bit cleaner.

Comment (NicolajNiels1) [reply src/components/molecules/AddTransaction.tsx]: Yes. Good catch 👍

File: src/mitochondria/index.ts (hunk 8 above)
Comment (NicolajNiels1) [reply src/mitochondria/index.ts]: Wonderful. Will fix!

File: src/mitochondria/transactions.ts (hunk 10 above)
Comment (NicolajNiels1) [reply src/mitochondria/transactions.ts]: Awesome! 💯 Will fix.

File: src/mitochondria/index.ts (hunk 8 above)
Comment (NicolajNiels1) [reply src/mitochondria/index.ts]: Fixed ✅

File: src/mitochondria/transactions.ts (hunk 10 above)
Comment (NicolajNiels1) [reply src/mitochondria/transactions.ts]: Fixed ✅

File: src/mitochondria/transactions.ts (hunk 11 above)
Comment (NicolajNiels1) [reply src/mitochondria/transactions.ts]: Of course! Thanks! 👍

File: src/mitochondria/transactions.ts (hunk 6 above)
Comment (FredrikAugust) [reply src/mitochondria/transactions.ts]: If you're thinking about the error handling, it is not in this case as catching and directly throwing an error is basically the same as doing nothing. The main difference is that it is (potentially) harder to see the true origin of the throw, as in this case it will be shown as coming from this function instead of its true source.

**Comment by NicolajNiels1** [comment]
I assigned @KaizerTilt 	so he can merge this after his (pending) review is approved 😊

File: src/mitochondria/transactions.ts [hunk 12]
```diff
@@ -1,29 +1,16 @@
...
  */
-export const createTransaction = async (transaction: INewTransaction, authState: IAuthState) => {
-  try {
-    const response = await post('/transaction', transaction, authState);
-    return response;
-  } catch (e) {
-    throw e;
-  }
+export const createTransaction = async (transaction: INewTransaction) => {
+    return await post('/transaction', transaction);
```
Comment (FredrikAugust) [inline src/mitochondria/transactions.ts]: If I understand this code correctly, then this does not handle expired/invalid tokens, right?

Comment (FredrikAugust) [reply src/mitochondria/transactions.ts]: I'll try to finish up and implement #73 so we have a generic `fetch` method called `fetchWithToken` that can be used for calls that need to be authorized.

File: src/components/organism/Balances.tsx [hunk 13]
```diff
@@ -12,7 +12,8 @@ interface IProps {
   className?: string;
 }
 
-const companyId = 1; // Hardcoded until we get a global company context.
+// tslint:disable-next-line: variable-name
```
Comment (KaizerTilt) [inline src/components/organism/Balances.tsx]: No need for snake case here I think

File: src/mitochondria/transactions.ts (hunk 12 above)
Comment (NicolajNiels1) [reply src/mitochondria/transactions.ts]: > I'll try to finish up and implement #73 so we have a generic `fetch` method called `fetchWithToken` that can be used for calls that need to be authorized.

@FredrikAugust 	Which use cases do we make API calls without authorization?
